observatory-dags/observatory/dags/telescopes/ucl_discovery.py

```diff
@@ -0,0 +1,314 @@
+
+    def transform(self):
+        """ Parse the csv file and for each eprint id store the relevant metadata in a dictionary and get the downloads
+        per country (between begin_date and end_date). The list of dictionaries is stored in a gzipped json lines file.
+        There might be multiple rows for 1 eprint id. Some columns only have a value in the first row, some columns
+        have values in multiple rows.
+
+        :return: None.
+        """
+        begin_date = self.start_date.strftime("%Y-%m-%d")
+        end_date = self.end_date.strftime("%Y-%m-%d")
+        total_downloads = 0
+        downloads_per_country = []
+        single_row_columns = {}
+        with open(self.download_path, 'r') as f:
+            csv_reader = csv.DictReader(f)
+
+            previous_id = None
+            results = []
+            multi_row_columns = {
+                'creators_name_family': [],
+                'creators_name_given': [],
+                'subjects': [],
+                'divisions': [],
+                'lyricists_name_family': [],
+                'lyricists_name_given': [],
+                'editors_name_family': [],
+                'editors_name_given': []
+            }
+            for row in csv_reader:
+                eprintid = row['eprintid']
+                # row with a new eprint id
+                if previous_id != eprintid:
+                    # add results of previous eprint id
+                    if previous_id:
+                        result = create_result_dict(begin_date, end_date, total_downloads, downloads_per_country,
+                                                    multi_row_columns, single_row_columns)
+                        results.append(result)
+                        for column in multi_row_columns:
+                            multi_row_columns[column] = []
+
+                    # store results of current eprint id
+                    single_row_columns = {
+                        'eprintid': row['eprintid'],
+                        'book_title': row['title'],
+                        'ispublished': row['ispublished'],
+                        'keywords': row['keywords'].split(', '),
+                        'abstract': row['abstract'],
+                        'date': row['date'],
+                        'publisher': row['publisher'],
+                        'official_url': row['official_url'],
+                        'oa_status': row['oa_status'],
+                        'language': row['language'],
+                        'doi': row['doi'],
+                        'isbn': row['isbn_13'],
+                        'language_elements': row['language_elements'],
+                        'series': row['series'],
+                        'pagerange': row['pagerange'],
+                        'pages': row['pages']
+                    }
+
+                    downloads_per_country, total_downloads = get_downloads_per_country(self.countries_url + eprintid)
```

Comment:
Since requesting downloads per country for each eprint id is a bottleneck, it would be good to experiment with the max_active_runs setting for the DAG to find a default value that increases the speed of the DAG (especially when backfilling) without making it unreliable.